controllers/backstage_controller.go

```diff
@@ -103,219 +93,146 @@ func (r *BackstageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	defer func(bs *bs.Backstage) {
 		if err := r.Client.Status().Update(ctx, bs); err != nil {
 			if errors.IsConflict(err) {
-				lg.V(1).Info("Backstage object modified, retry reconciliation", "Backstage Object", bs)
+				lg.V(1).Info("Backstage object modified, retry syncing status", "Backstage Object", bs)
 				return
 			}
 			lg.Error(err, "Error updating the Backstage resource status", "Backstage Object", bs)
 		}
 	}(&backstage)
 
 	if len(backstage.Status.Conditions) == 0 {
-		setStatusCondition(&backstage, bs.ConditionDeployed, v1.ConditionFalse, bs.DeployInProgress, "Deployment process started")
+		setStatusCondition(&backstage, bs.BackstageConditionTypeDeployed, metav1.ConditionFalse, bs.BackstageConditionReasonInProgress, "Deployment process started")
 	}
 
-	if pointer.BoolDeref(backstage.Spec.Database.EnableLocalDb, true) {
-
-		/* We use default strogeclass currently, and no PV is needed in that case.
-		If we decide later on to support user provided storageclass we can enable pv creation.
-		if err := r.applyPV(ctx, backstage, req.Namespace); err != nil {
-			return ctrl.Result{}, err
-		}
-		*/
-
-		err := r.reconcileLocalDbStatefulSet(ctx, &backstage, req.Namespace)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-
-		err = r.reconcileLocalDbServices(ctx, &backstage, req.Namespace)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-	} else { // Clean up the deployed local db resources if any
-		if err := r.cleanupLocalDbResources(ctx, backstage); err != nil {
-			setStatusCondition(&backstage, bs.ConditionDeployed, v1.ConditionFalse, bs.DeployFailed, fmt.Sprintf("failed to delete Database Services:%s", err.Error()))
-			return ctrl.Result{}, fmt.Errorf("failed to delete Database Service: %w", err)
-		}
+	// 1. Preliminary read and prepare external config objects from the specs (configMaps, Secrets)
+	// 2. Make some validation to fail fast
+	spec, err := r.preprocessSpec(ctx, backstage)
+	if err != nil {
+		setStatusCondition(&backstage, bs.BackstageConditionTypeDeployed, metav1.ConditionFalse, bs.BackstageConditionReasonFailed, fmt.Sprintf("failed to preprocess backstage spec %s", err))
+		return ctrl.Result{}, fmt.Errorf("failed to preprocess backstage spec %w", err)
 	}
 
-	err := r.reconcileBackstageDeployment(ctx, &backstage, req.Namespace)
+	// This creates array of model objects to be reconsiled
+	bsModel, err := model.InitObjects(ctx, backstage, spec, r.OwnsRuntime, r.IsOpenShift, r.Scheme)
 	if err != nil {
-		return ctrl.Result{}, err
+		setStatusCondition(&backstage, bs.BackstageConditionTypeDeployed, metav1.ConditionFalse, bs.BackstageConditionReasonFailed, fmt.Sprintf("failed to initialize backstage model %s", err))
+		return ctrl.Result{}, fmt.Errorf("failed to initialize backstage model %w", err)
 	}
 
-	if err := r.reconcileBackstageService(ctx, &backstage, req.Namespace); err != nil {
-		return ctrl.Result{}, err
+	err = r.applyObjects(ctx, bsModel.Objects)
+	if err != nil {
```

Comment:
Errors of conflicting objects are temporary. They will be cleared in next reconciliations and should not cause the CR status to get updated.

Reply:
This kind of conflicting errors are vey common in operators and all they need to do is to retry the reconciliation. They are typically self-recoverable so they should not cause un-necessary status updates.